From bc1a5f7faaec840165d9e367cb29737a2ab0692b Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Date: Thu, 10 Dec 2020 14:08:40 +0200
Subject: [PATCH 08/78] rtc: s32gen1: Add sentinel to device ID list

This fixes an issue reported by KASAN:

==================================================================
BUG: KASAN: global-out-of-bounds in __of_match_node+0x70/0xb8
Read of size 1 at addr ffffffd010bce068 by task swapper/0/1

CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.4.69-rt39-00007-ga45ded9 #742
Hardware name: Freescale S32G274 (DT)
Call trace:
 dump_backtrace+0x0/0x200
 show_stack+0x14/0x20
 dump_stack+0xd8/0x144
 print_address_description.isra.7+0x134/0x360
 __kasan_report+0x124/0x220
 kasan_report+0xc/0x18
 __asan_load1+0x60/0x68
 __of_match_node+0x70/0xb8
 of_match_node+0x38/0x60
 of_match_device+0x30/0x50
 platform_match+0x74/0x120
 __driver_attach+0x40/0x120
 bus_for_each_dev+0xdc/0x150
 driver_attach+0x30/0x40
 bus_add_driver+0x238/0x2d8
 driver_register+0xb4/0x1c8
 __platform_driver_register+0x7c/0x88
 s32gen1_rtc_driver_init+0x18/0x20
 do_one_initcall+0xac/0x25c
 kernel_init_freeable+0x1cc/0x274
 kernel_init+0x10/0x118
 ret_from_fork+0x10/0x18

The buggy address belongs to the variable:
 s32gen1_rtc_of_match+0xc8/0x600

Memory state around the buggy address:
 ffffffd010bcdf00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa
 ffffffd010bcdf80: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00 00
>ffffffd010bce000: 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa fa
                                                          ^
 ffffffd010bce080: fa fa fa fa 00 00 00 00 01 fa fa fa fa fa fa fa
 ffffffd010bce100: 04 fa fa fa fa fa fa fa 07 fa fa fa fa fa fa fa
==================================================================

Issue: ALB-6156
Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
---
 drivers/rtc/rtc-s32gen1.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/rtc/rtc-s32gen1.c b/drivers/rtc/rtc-s32gen1.c
index 7d7abeeb345d..23120728bcc2 100644
--- a/drivers/rtc/rtc-s32gen1.c
+++ b/drivers/rtc/rtc-s32gen1.c
@@ -617,6 +617,7 @@ static int s32gen1_rtc_resume(struct device *dev)
 
 static const struct of_device_id s32gen1_rtc_of_match[] = {
 	{.compatible = "fsl,s32gen1-rtc" },
+	{ },
 };
 
 static SIMPLE_DEV_PM_OPS(s32gen1_rtc_pm_ops,
-- 
2.25.1

